### PR TITLE
Introducing new timestepper IterationAdaptivePostprocessorDT

### PR DESCRIPTION
**Pull request by giopastor**
### Complete each of the following items before creating a PR
- Include any relevant design information for your change
  The new Timestepper (IterationAdaptivePostprocessorDT) extends IterationAdaptiveDT to include the time step from a postprocessor as an additional criterion. It combines the capabilities of IterationAdaptiveDT and PostprocessorDT. It is thought for BISON applications wherein we need to allow for material-model based time step criteria.
  The new routine is based on IterationAdaptiveDT and retains all capabilities of IterationAdaptiveDT. The main differences are
- It inherits also from PostprocessorInterface, in order to be able to retrieve the postprocessor values (variable _pps_value).
- It includes a new method, limitDTByPostprocessor, which is called from constrainStep. This method makes sure the time step does not exceed the postprocessor value.
- Follow [Coding Standards](http://mooseframework.org/wiki/CodeStandards/)
- Submit one or more [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/) or modify existing test cases
  Done
- Reference or close a specific issue (refs # or closes #)
  Refs #6576

### CODE REVIEW

File: framework/src/timesteppers/IterationAdaptivePostprocessorDT.C [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+    _timestep_limiting_function = &_fe_problem.getFunction(getParam<FunctionName>("timestep_limiting_function"), isParamValid("_tid") ? getParam<THREAD_ID>("_tid") : 0);
+    _piecewise_timestep_limiting_function = dynamic_cast<Piecewise*>(_timestep_limiting_function);
+    if (_piecewise_timestep_limiting_function)
+    {
+      unsigned int time_size = _piecewise_timestep_limiting_function->functionSize();
+      _times.resize(time_size);
+      for (unsigned int i=0; i < time_size; ++i)
+      {
+       _times[i] = _piecewise_timestep_limiting_function->domain(i);
+      }
```
Comment (permcody) [inline framework/src/timesteppers/IterationAdaptivePostprocessorDT.C]: We discourage extraneous braces. They appear throughout your code. If you only have a single statement. Please drop the brackets

File: framework/src/timesteppers/IterationAdaptivePostprocessorDT.C [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+{
+  if (isParamValid("timestep_limiting_function"))
+  {
+    _timestep_limiting_function = &_fe_problem.getFunction(getParam<FunctionName>("timestep_limiting_function"), isParamValid("_tid") ? getParam<THREAD_ID>("_tid") : 0);
+    _piecewise_timestep_limiting_function = dynamic_cast<Piecewise*>(_timestep_limiting_function);
+    if (_piecewise_timestep_limiting_function)
+    {
+      unsigned int time_size = _piecewise_timestep_limiting_function->functionSize();
+      _times.resize(time_size);
+      for (unsigned int i=0; i < time_size; ++i)
```
Comment (permcody) [inline framework/src/timesteppers/IterationAdaptivePostprocessorDT.C]: Spacing around binary operators `i = 0`

File: framework/src/timesteppers/IterationAdaptivePostprocessorDT.C [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+        newValue = _timestep_limiting_function->value(_time_old+limitedDT,dummyPoint);
+        change = std::abs(newValue-oldValue);
+      }
+      while (change > _max_function_change);
+    }
+  }
+  _at_function_point = false;
+  if (_piecewise_timestep_limiting_function && _force_step_every_function_point)
+  {
+    for (unsigned int i=0; i<_times.size()-1; ++i)
```
Comment (permcody) [inline framework/src/timesteppers/IterationAdaptivePostprocessorDT.C]: spacing

File: framework/src/timesteppers/IterationAdaptivePostprocessorDT.C [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+  if (_verbose)
+    _console << diag.str();
+}
+
+void
+IterationAdaptivePostprocessorDT::limitDTByPostprocessor(Real & limitedDT)
+{
+  std::ostringstream diag;
+
+  if( _pps_value > 0. && limitedDT > _pps_value)
```
Comment (permcody) [inline framework/src/timesteppers/IterationAdaptivePostprocessorDT.C]: spacing after the `if`. This is probably why your pre-check failed.

File: framework/src/timesteppers/IterationAdaptivePostprocessorDT.C [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+  return params;
+}
+
+IterationAdaptivePostprocessorDT::IterationAdaptivePostprocessorDT(const InputParameters & parameters) :
+  TimeStepper(parameters),
+  PostprocessorInterface(parameters),
+  _dt_old(declareRestartableData<Real>("dt_old",0.0)),
+  _input_dt(getParam<Real>("dt")),
+  _tfunc_last_step(declareRestartableData<bool>("tfunc_last_step",false)),
+  _sync_last_step(declareRestartableData<bool>("sync_last_step",false)),
```
Comment (permcody) [inline framework/src/timesteppers/IterationAdaptivePostprocessorDT.C]: Any restartable data type needs to be declared as a reference. I see several here that are not references in your header file. These won't work correctly with restart which makes me think you don't need them marked as restartable.

File: framework/src/timesteppers/IterationAdaptivePostprocessorDT.C [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+{
+  Real dt(_dt_old);
+  std::ostringstream diag;
+
+  if ( _cutback_occurred ) //Don't allow it to grow this step, but shrink if needed
+  {
+    _cutback_occurred = false;
+    if (_adaptive_timestepping)
+    {
+      bool allowToGrow(false);
```
Comment (permcody) [inline framework/src/timesteppers/IterationAdaptivePostprocessorDT.C]: prefer `=` initialization of plain old data types over `()` initialization

**Comment by bwspenc** [comment]
@giopastor It seems to me that we shouldn't create a separate Timestepper just to add the additional feature of being able to control the time step with a postprocessor.  IterationAdaptiveDT already has a few options to do things like limiting the timestep to limit the change in a function. It seems like it would be reasonable to add this as just another option to IterationAdaptiveDT.

**Comment by dschwen** [comment]
You can use a postprocessor value in a function. It seems to mat that no
change at all is needed here.

**Comment by giopastor** [comment]
@permcody @bwspenc Thanks a lot for the comments. Ben, I was thinking the same. The reason why I originally did a new postprocessor is that this one features a multiple inheritance (TimeStepper, PostprocessorInterface), differently from IterationAdaptiveDT. If it is OK, I will just extend IterationAdaptiveDT to make it inherit also from PostprocessorInterface and include the new capability.

**Comment by giopastor** [comment]
@dschwen I know a postprocessor can return a function value, is the other way around possible?

**Comment by permcody** [comment]
Inheriting from multiple interfaces shouldn't be a problem. We do it all
the time in MOOSE and we don't consider that multiple inheritance because
we don't store data in interfaces (with the exceptions of parameter data).
As long as you avoid inheriting from multiple "concrete" (as opposed to
interfaces) you should be fine.

**Comment by giopastor** [comment]
@permcody Oh OK thanks! Will do then

**Comment by dschwen** [comment]
@giopastor , look at the `vars` and `vals` parameters to `ParsedFunction`. Example:

```
[Functions]
  [./func]
    type = ParsedFunction
    value = p
    vars = p
    vals = mypp
  [../]
[]

[Postprocessors]
  [./mypp]
    type = AwesomePostprocessor
  [../]
[]
```

**Comment by permcody** [comment]
Closing based on comments posted here.

**Pull request closed without merging**